### ISSUE TITLE
Support LSP servers that return locations within archive files

### ISSUE THREAD

**Issue by tapayne88** [neovim/neovim#14867]
In yarn v2 Plug'n'Play projects typescript's language server (https://github.com/theia-ide/typescript-language-server) emits files with `.zip` extensions, paths like
```
/Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip/node_modules/@testing-library/react/types/index.d.ts
```

There is some thinking that the language server should handle recognising these archive files and emit neovim friendly paths prefixed with `zipfile:` and converting `.zip/` to `.zip::`. As these LSPs are used by editors other than neovim it feels like the wrong place to fix the issue.

It is possible to register a custom handler for the typescript language server which can fix the problem but this will only work if you're using neovim's builtin `vim.lsp.buf.definition `, it won't help LSP aware plugins like telescope's `lsp_definitions`.

A possible solution I thought might work would be to patch `jump_to_location` to recognise these archive files and convert the paths to neovim friendly ones but appreciate I am no expert in this area!

What do you think?

**Comment by mjlbach** [comment]
I'm generally opposed to adding any language server specific workarounds to the core client, especially as this seems like a niche case. One of the advantages of the design of the core client is that it's incredibly easy to override any of the handlers, so it should be trivial to create a plugin (or modify something like https://github.com/jose-elias-alvarez/nvim-lsp-ts-utils) to support this functionality. 

WDYT @folke & @mfussenegger?

**Comment by tapayne88** [comment]
I agree it does feel a bit too specific to be in core but couldn't figure out a way to get handlers between the LSP and other plugins like telescope.

Perhaps we could expose functions that request LSP actions and include any registered custom handlers?

**Comment by folke** [comment]
As far as I know only yarn is using this. Last time I checked (probably a year ago), vscode didnt event support those uris for got to definition etc. Might have been included in vscode in the meantime.

It would definitely make sense for tsserver to at least emit proper uris using something like `zipfile:`.

With those uris, it should be possible to implement a plugin that correctly extracts the files needed to display in the buffer.

@mjlbach is there currently any way to hook into specific uri schemes?

What I mean is, at this moment, if you have an uri with a custom scheme that is opened in a buffer, nothing happens as far as I know, so is there a way to `register` handlers for specific uri schemas?

**Comment by folke** [comment]
A `BufReadCmd` auto command can be used to load the actual file from the zip file in the buffer.

Could be an enhancement to @jose-elias-alvarez plugin like @mjlbach mentioned :)

**Comment by tapayne88** [comment]
Yeah, [vim-rzip](https://github.com/lbrayner/vim-rzip) can handle those filenames however tsserver would emit something like
```
/Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip/node_modules/@testing-library/react/types/index.d.ts
```
and something would need to convert it to
```
zipfile:/Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip::node_modules/@testing-library/react/types/index.d.ts
```
Yarn does have a layer which can do this ([how it could work](https://github.com/tapayne88/berry/commit/ed189a9cbd97fe94e066b0a10e3b8c56d3c2b97d)) however the language server needs to ignore the non-standard URIs.

It feels unfair for the language server to [have handling to ignore non-standard URIs](https://github.com/theia-ide/typescript-language-server/pull/220/files#r654288222) as it's a specific yarn / neovim requirement.

It would be nice if something on the neovim side could recognise the `.zip` extension and produce the correct URI. This is my current handler but it doesn't work with telescope
```lua
handlers = {
    ["textDocument/definition"] = function(_, method, result)
        local patched_paths = vim.tbl_map(function(entry)
            local uri = entry.uri
            local is_zip = uri:match(".zip/")
            if is_zip ~= nil then
                uri = uri:gsub("^(file://)", "%1zipfile:"):gsub(".zip/",
                                                                ".zip::")
            end

            return vim.tbl_extend("force", entry, {uri = uri})
        end, result)
        -- runtime/lua/vim/lsp/handlers.lua
        location_handler(_, method, patched_paths)
    end
}
```

**Comment by folke** [comment]
Shouldn't Yarn return proper uris instead of those file like paths?

**Comment by tapayne88** [comment]
Yeah probably 😄 
Forgive my ignorance but would this qualify as a valid URI, specifically the `::` part?
```
zipfile:/Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip::node_modules/@testing-library/react/types/index.d.ts
```

**Comment by folke** [comment]
I believe you should have triple slashes after the colon, so:

```
zipfile:///Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip::node_modules/@testing-library/react/types/index.d.ts
```

**Comment by tapayne88** [comment]
I _think_ ([according to wikipedia - Syntax section](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax)) if we don't have an authority in the URI it shouldn't begin with `//` but there seems to be an exception for `file` schemes 🤷 

Unfortunately it seems neovim requires URI's path to have the prefix `//`
https://github.com/neovim/neovim/blob/b585f723bcbaa10a091fce5a61659b331b1467b9/runtime/lua/vim/uri.lua#L77

**Comment by tapayne88** [comment]
So I can get URI of a valid pattern to be passed all the way up from tsserver to neovim core, paths like this
```
zipfile:/Users/thomas.payne/git/yarn-pnp-demo/.yarn/cache/@testing-library-react-npm-11.2.7-3a0469c756-389c9f3e83.zip::node_modules/@testing-library/react/types/index.d.ts
```

However there is an issue when they arrive in core as the scheme doesn't pass the regex ([see above](https://github.com/neovim/neovim/issues/14867#issuecomment-864858800)). The path to the zip file has also been URI encoded and needs decoding similar to how `file://` schemes are handled.
https://github.com/neovim/neovim/blob/b585f723bcbaa10a091fce5a61659b331b1467b9/runtime/lua/vim/uri.lua#L100

I _think_ a change like this would resolve these issues and allow yarn PnP files to be opened correctly.
https://github.com/tapayne88/neovim/commit/424f0b6409a95befd3b66adb1a22c1d96e5c90fb

What do you think?

**Comment by mfussenegger** [comment]
I think fixing the regex to account for the missing authority / optional `//` sounds good.

> I think a change like this would resolve these issues and allow yarn PnP files to be opened correctly.
tapayne88@424f0b6

The URI code probably shouldn't be aware of any specific scheme's - with the exception of `file`.

**Comment by tapayne88** [comment]
> The URI code probably shouldn't be aware of any specific scheme's - with the exception of `file`.

@mfussenegger do you think it would be appropriate to `uri_decode` all URIs not just `file` schemes? `zipfile` schemes are just filepaths with a different scheme to trigger [vim-rzip's](https://github.com/lbrayner/vim-rzip) autocommands so need to be decoded

**Comment by mfussenegger** [comment]
> > The URI code probably shouldn't be aware of any specific scheme's - with the exception of `file`.
> 
> @mfussenegger do you think it would be appropriate to `uri_decode` all URIs not just `file` schemes? `zipfile` schemes are just filepaths with a different scheme to trigger [vim-rzip's](https://github.com/lbrayner/vim-rzip) autocommands so need to be decoded

decoding all URIs would break other use-cases, see https://github.com/neovim/neovim/pull/12351 

Not sure how to handle this.

**Comment by folke** [comment]
@tapanprakasht you mean the structure of the uri is different than the one supported by vim-rzip?

You could also make Yarn return uris like `yarnpnp:...`. Then have an `BufReadCmd yarnpnp:/*` the converts the uri to the one supported by vim-rzip and does `do BufReadCmd zipfile....`.

Would that work?

**Comment by folke** [comment]
I agree with @mfussenegger that Neovim should only do special handling for `file://` uris. Any other scheme should be handled by a plugin.

**Comment by folke** [comment]
Just for reference:
* [docs for uris](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#uri) in the LSP spec
* [uri parser](https://github.com/Microsoft/vscode-uri) used by vscode to handle uris within lsp

**Comment by tapayne88** [comment]
Yeah, sounds good, thanks for your help @folke & @mfussenegger!

Just to confirm we think updating the `URI_SCHEME_PATTERN` regex to allow `//` to be optional is good but leave it at that? Do you think there is any utility in exporting the `uri_decode` function? Doesn't seem too bad to just copy/paste

**Comment by tapayne88** [comment]
The required change proposed in #14959 was implemented in #15017 so I don't think there are any blockers to supporting archive file locations with LSP. It's possible with TypeScript in yarn v2 Plug'n'Play projects.

- https://github.com/neovim/nvim-lspconfig/pull/1078
- https://github.com/theia-ide/typescript-language-server/pull/220
- https://github.com/yarnpkg/berry/pull/3041

For info on how to open nested archive locations in vim see https://github.com/lbrayner/vim-rzip/issues/15
